### ISSUE TITLE
AlertManager is ignoring Discord's Message Limits

### ISSUE THREAD

**Issue by siegenthalerroger** [prometheus/alertmanager#3310]
**What did you do?**

Set up alertmanager with the prometheus-operator, setup the kube-prometheus library of metrics and alerts, configured some routes and a few discord and slack receivers.

**What did you expect to see?**

The same alerts arriving in slack and in discord.

**What did you see instead? Under which circumstances?**

I did not receive all messages that were sent to slack arriving in discord. This is the same behaviour that was seen using a discord-bridge previously, so nothing new. However this time with the native integration, I can actually see some "useful" logs.

```
ts=2023-03-30T12:32:21.595Z caller=dispatch.go:352 level=error component=dispatcher msg="Notify for alerts failed" num_alerts=20 err="elder-discord-minor/discord[0]: notify retry canceled due to unrecoverable error after 1 attempts: unexpected status code 400: {\"embeds\": [\"0\"]}"
```

A lot of googling about potential rate-limits or formatting issues, etc. I discovered the following stackoverflow answer that contains a list of limits that discord webhooks have. These apparently result in the above status code error and would definitely be possible options. Make sure to look at the "coming from a search engine" part of the first answer, the beginning isn't exactly relevant.
- https://stackoverflow.com/questions/53935198/in-my-discord-webhook-i-am-getting-the-error-embeds-0#:~:text=For%20anyone%20else%20coming%20from%20search%20engine

**Environment**

* System information:

```
Linux 5.15.0-1034-azure x86_64
```

* Alertmanager version:

```
alertmanager, version 0.25.0 (branch: HEAD, revision: 258fab7cdd551f2cf251ed0348f0ad7289aee789)
  build user:       root@abe866dd5717
  build date:       20221222-14:51:36
  go version:       go1.19.4
  platform:         linux/amd64
```

* Prometheus version:

```
prometheus, version 2.37.6 (branch: HEAD, revision: 8ade24a23af6be0f35414d6e8ce09598446c29a2)
  build user:       root@5f96027a7c3e
  build date:       20230220-09:36:40
  go version:       go1.19.6
  platform:         linux/amd64
```

* Alertmanager configuration file:
```
global:
  resolve_timeout: 5m # has no effect on prometheus originated alerts

templates:
- '*.tmpl'

inhibit_rules:
  # inhibit less severe alerts of same name (e.g. if multiple levels can be triggered)
  - equal:
    - namespace
    - alertname
    source_matchers:
    - severity = critical
    target_matchers:
    - severity =~ warning|info
  # inhibit less severe alerts of same name (e.g. if multiple levels can be triggered)
  - equal:
    - namespace
    - alertname
    source_matchers:
    - severity = warning
    target_matchers:
    - severity = info

route:
  group_wait: 15s
  group_interval: 5m
  group_by: 
    - vendor
    - severity
    - env
    - app
  continue: false
  routes:
    - matchers: 
        - vendor =~ .*elder.*
      continue: false
      routes:
        - receiver: 'elder-discord-critical'
          matchers:
            - env = prod
            - severity = critical
          repeat_interval: 2h
<...>
        - receiver: 'elder-discord-minor'
          matchers:
            - env =~ ^(prod|qa)$
            - severity = minor
          repeat_interval: 12h
        - receiver: 'dummy-receiver' # remove all watchdog alerts (not monitored with current setup)
          matchers:
            - alertname = Watchdog
        - receiver: 'dummy-receiver' # route info inhibitors to a null receiver
          matchers:
            - alertname = InfoInhibitor
<...>

receivers:
  - name: 'dummy-receiver'
 <...>
  - name: 'elder-discord-minor'
    slack_configs:
      - api_url: <secret>
        channel: alert-minor
        send_resolved: true
    discord_configs:
      - webhook_url: <secret>
        send_resolved: true
<...>
```

* Logs:
```
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[3ceb64c][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[00dc451][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[f597fb1][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[cdaf1d9][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[098f72f][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[480d0e7][active]
ts=2023-03-30T12:47:21.053Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[f087a3d][active]
ts=2023-03-30T12:47:21.054Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[9b8f3a5][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[f597fb1][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[00dc451][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[3ceb64c][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[527d2b2][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[f087a3d][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[098f72f][active]
ts=2023-03-30T12:47:21.055Z caller=dispatch.go:163 level=debug component=dispatcher msg="Received alert" alert=TooLargeConsumerGroupLag[9b8f3a5][active]
ts=2023-03-30T12:47:21.427Z caller=dispatch.go:515 level=debug component=dispatcher aggrGroup="{}/{vendor=~\".*elder.*\"}/{env=~\"^(prod|qa)$\",severity=\"minor\"}:{app=\"kafka\", env=\"prod\", severity=\"minor\", vendor=\"elderbyte\"}" msg=flushing alerts="[TooLargeConsumerGroupLag[f597fb1][active] TooLargeConsumerGroupLag[811cb49][active] TooLargeConsumerGroupLag[75d6d0b][active] TooLargeConsumerGroupLag[527d2b2][active] TooLargeConsumerGroupLag[8490b24][active] TooLargeConsumerGroupLag[3ceb64c][active] TooLargeConsumerGroupLag[3083125][active] TooLargeConsumerGroupLag[f087a3d][active] TooLargeConsumerGroupLag[9b8f3a5][active] TooLargeConsumerGroupLag[04da413][resolved] TooLargeConsumerGroupLag[1381272][active] TooLargeConsumerGroupLag[37c260c][active] TooLargeConsumerGroupLag[45db3d3][active] TooLargeConsumerGroupLag[00dc451][active] TooLargeConsumerGroupLag[8bf9279][active] TooLargeConsumerGroupLag[098f72f][active] TooLargeConsumerGroupLag[9b18365][active] TooLargeConsumerGroupLag[480d0e7][active] TooLargeConsumerGroupLag[cdaf1d9][active] TooLargeConsumerGroupLag[3bc593f][resolved]]"
ts=2023-03-30T12:47:21.427Z caller=discord.go:84 level=debug integration=discord incident="{}/{vendor=~\".*elder.*\"}/{env=~\"^(prod|qa)$\",severity=\"minor\"}:{app=\"kafka\", env=\"prod\", severity=\"minor\", vendor=\"elderbyte\"}"
```

**Comment by siegenthalerroger** [comment]
As the webhook call in question has >10 alerts that are being flushed in one, I assume the issue is with discord's limit about the number of embeds per message (10).

**Comment by simonpasquier** [comment]
For reference, these are the documented limits:
https://discord.com/developers/docs/resources/channel#embed-object-embed-limits
https://discord.com/developers/docs/resources/webhook#execute-webhook-jsonform-params

Alertmanager would need to truncate values exceeding these limits.

**Comment by guitcastro** [comment]
It's possible to avoid messages being flushed into one?

**Comment by serejke** [comment]
As Simon mentioned above, the embeds have a `title` [limit](https://discord.com/developers/docs/resources/channel#embed-object-embed-limits) of 256 characters, which with the [default title template](https://github.com/prometheus/alertmanager/blob/3b9cd0408e682c88b57d9398bcb3402353948632/config/notifiers.go#L51) is too restrictive.

I believe the Discord integration should be responsible for truncating the title/body with a shorter template if the rendered `Title` / `Description` exceed the limit. 

https://github.com/prometheus/alertmanager/blob/550952eb51be17ab9cb5c45b57191efd9a0da8cb/notify/discord/discord.go#L113

Otherwise the alerts will never reach the Discord with the error:
> discord[0]: notify retry canceled due to unrecoverable error after 1 attempts: unexpected status code 400: {\"embeds\": [\"0\"]}

From my knowledge, the status 400 is not related to _rate limit_ (429), which is _another_ issue of the Discord integration: on receiving 429 response, it retries immediately, without waiting for the delay specified in header. Retrier is configured with empty policy for handling 429:
https://github.com/prometheus/alertmanager/blob/550952eb51be17ab9cb5c45b57191efd9a0da8cb/notify/discord/discord.go#L60
this can be improved.

### Workaround for the message limit
I had to override the `title` and `message` templates for our use case. This allows to pass message limits:

`templates.tmpl`
```
{{ define "our.discord.title" }}
    Fire! {{ .GroupLabels.SortedPairs.Values | join " " }}
{{ end }}

{{ define "our.discord.message" }}
    {{ range .Alerts.Firing }}
      Alert: **{{ printf "%.150s" .Annotations.summary }}** ({{ .Labels.severity }})
      Description: {{ printf "%.150s" .Annotations.description }}
      Alertname: {{ .Labels.alertname }}
      Namespace: {{ .Labels.namespace }}
      Service: {{ .Labels.service }}
    {{ end }}

    {{ if gt (len .Alerts.Resolved) 0 }}
        Also {{ .Alerts.Resolved | len }} resolved alerts.
    {{ end }}
{{ end }}
```

then specify the template `alertmanager.yaml`:
```
receivers:
  - name: discord
    discord_configs:
      - webhook_url: 'https://discord.com/api/webhooks/<WEB_HOOK>/<WEB_HOOK>'
        message: '{{ template "our.discord.message" . }}'
        title: '{{ template "our.discord.title" . }}'
```

**Comment by metalmatze** [comment]
We just happened to have run into the same issues at Polar Signals now. Given I did the original implementation I'm and we're facing the same problem, I'm happy to take a look at how we can fix this issue.

**Comment by metalmatze** [comment]
Thanks, everyone for the great discussion so far. That should be a great foundation for improving the current implementation.

**Comment by simonpasquier** [comment]
> From my knowledge, the status 400 is not related to _rate limit_ (429), which is _another_ issue of the Discord integration: on receiving 429 response, it retries immediately, without waiting for the delay specified in header. Retrier is configured with empty policy for handling 429:

@serejke  I've added Discord to https://github.com/prometheus/alertmanager/issues/2205

**Comment by finaldie** [comment]
Tried the new alertmanager verion `v0.27.0-rc.0` which included the fixes, and works out of the box. :+1:

**Comment by ashwaq06** [comment]
> receivers:
>   - name: discord
>     discord_configs:
>       - webhook_url: 'https://discord.com/api/webhooks/<WEB_HOOK>/<WEB_HOOK>'
>         message: '{{ template "our.discord.message" . }}'
>         title: '{{ template "our.discord.title" . }}'

I've attempted several methods to templatize the alerts while using the `prometheus-stack` Helm chart. Specifically, I've tried creating templates through a `ConfigMap` and volume mounting it, but those attempts haven't worked. Additionally, I'm encountering same issue with configuring Discord webhooks for alert notifications. Can anyone provide guidance on how to properly set this up when using Alertmanager through the `prometheus-stack` Helm chart? 
Any help or suggestions would be greatly appreciated
